File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceBranchService.java
/*
 * Copyright 2021 Apollo Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package com.ctrip.framework.apollo.portal.service;

import com.ctrip.framework.apollo.common.dto.*;
import com.ctrip.framework.apollo.common.exception.BadRequestException;
import com.ctrip.framework.apollo.portal.api.AdminServiceAPI;
import com.ctrip.framework.apollo.portal.component.ItemsComparator;
import com.ctrip.framework.apollo.portal.constant.TracerEventType;
import com.ctrip.framework.apollo.portal.entity.bo.NamespaceBO;
import com.ctrip.framework.apollo.portal.environment.Env;
import com.ctrip.framework.apollo.portal.spi.UserInfoHolder;
import com.ctrip.framework.apollo.tracer.Tracer;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collections;
import java.util.List;

@Service
public class NamespaceBranchService {

    private final ItemsComparator itemsComparator;
    private final UserInfoHolder userInfoHolder;
    private final NamespaceService namespaceService;
    private final ItemService itemService;
    private final AdminServiceAPI.NamespaceBranchAPI namespaceBranchAPI;
    private final ReleaseService releaseService;

    public NamespaceBranchService(
            final ItemsComparator itemsComparator,
            final UserInfoHolder userInfoHolder,
            final NamespaceService namespaceService,
            final ItemService itemService,
            final AdminServiceAPI.NamespaceBranchAPI namespaceBranchAPI,
            final ReleaseService releaseService) {
        this.itemsComparator = itemsComparator;
        this.userInfoHolder = userInfoHolder;
        this.namespaceService = namespaceService;
        this.itemService = itemService;
        this.namespaceBranchAPI = namespaceBranchAPI;
        this.releaseService = releaseService;
    }


    @Transactional
    public NamespaceDTO createBranch(String appId, Env env, String parentClusterName, String namespaceName) {
        String operator = userInfoHolder.getUser().getUserId();
        return createBranch(appId, env, parentClusterName, namespaceName, operator);
    }

    @Transactional
    public NamespaceDTO createBranch(String appId, Env env, String parentClusterName, String namespaceName, String operator) {
        NamespaceDTO createdBranch = namespaceBranchAPI.createBranch(appId, env, parentClusterName, namespaceName,
                operator);

        Tracer.logEvent(TracerEventType.CREATE_GRAY_RELEASE, String.format("%s+%s+%s+%s", appId, env, parentClusterName,
                namespaceName));
        return createdBranch;

    }

    public GrayReleaseRuleDTO findBranchGrayRules(String appId, Env env, String clusterName,
                                                  String namespaceName, String branchName) {
        return namespaceBranchAPI.findBranchGrayRules(appId, env, clusterName, namespaceName, branchName);

    }

    public void updateBranchGrayRules(String appId, Env env, String clusterName, String namespaceName,
                                      String branchName, GrayReleaseRuleDTO rules) {

        String operator = userInfoHolder.getUser().getUserId();
        updateBranchGrayRules(appId, env, clusterName, namespaceName, branchName, rules, operator);
    }

    public void updateBranchGrayRules(String appId, Env env, String clusterName, String namespaceName,
                                      String branchName, GrayReleaseRuleDTO rules, String operator) {
        rules.setDataChangeCreatedBy(operator);
        rules.setDataChangeLastModifiedBy(operator);

        namespaceBranchAPI.updateBranchGrayRules(appId, env, clusterName, namespaceName, branchName, rules);

        Tracer.logEvent(TracerEventType.UPDATE_GRAY_RELEASE_RULE,
                String.format("%s+%s+%s+%s", appId, env, clusterName, namespaceName));
    }

    public void deleteBranch(String appId, Env env, String clusterName, String namespaceName,
                             String branchName) {

        String operator = userInfoHolder.getUser().getUserId();
        deleteBranch(appId, env, clusterName, namespaceName, branchName, operator);
    }

    public void deleteBranch(String appId, Env env, String clusterName, String namespaceName,
                             String branchName, String operator) {
        namespaceBranchAPI.deleteBranch(appId, env, clusterName, namespaceName, branchName, operator);

        Tracer.logEvent(TracerEventType.DELETE_GRAY_RELEASE,
                String.format("%s+%s+%s+%s", appId, env, clusterName, namespaceName));
    }


    public ReleaseDTO merge(String appId, Env env, String clusterName, String namespaceName,
                            String branchName, String title, String comment,
                            boolean isEmergencyPublish, boolean deleteBranch) {
        String operator = userInfoHolder.getUser().getUserId();
        return merge(appId, env, clusterName, namespaceName, branchName, title, comment, isEmergencyPublish, deleteBranch, operator);
    }

    public ReleaseDTO merge(String appId, Env env, String clusterName, String namespaceName,
                            String branchName, String title, String comment,
                            boolean isEmergencyPublish, boolean deleteBranch, String operator) {

        ItemChangeSets changeSets = calculateBranchChangeSet(appId, env, clusterName, namespaceName, branchName, operator);

        ReleaseDTO mergedResult =
                releaseService.updateAndPublish(appId, env, clusterName, namespaceName, title, comment,
                        branchName, isEmergencyPublish, deleteBranch, changeSets);

        Tracer.logEvent(TracerEventType.MERGE_GRAY_RELEASE,
                String.format("%s+%s+%s+%s", appId, env, clusterName, namespaceName));

        return mergedResult;
    }

    private ItemChangeSets calculateBranchChangeSet(String appId, Env env, String clusterName, String namespaceName,
                                                    String branchName, String operator) {
        NamespaceBO parentNamespace = namespaceService.loadNamespaceBO(appId, env, clusterName, namespaceName);

        if (parentNamespace == null) {
            throw new BadRequestException("base namespace not existed");
        }

        if (parentNamespace.getItemModifiedCnt() > 0) {
            throw new BadRequestException("Merge operation failed. Because master has modified items");
        }

        List<ItemDTO> masterItems = itemService.findItems(appId, env, clusterName, namespaceName);

        List<ItemDTO> branchItems = itemService.findItems(appId, env, branchName, namespaceName);

        ItemChangeSets changeSets = itemsComparator.compareIgnoreBlankAndCommentItem(parentNamespace.getBaseInfo().getId(),
                masterItems, branchItems);
        changeSets.setDeleteItems(Collections.emptyList());
        changeSets.setDataChangeLastModifiedBy(operator);
        return changeSets;
    }

    public NamespaceDTO findBranchBaseInfo(String appId, Env env, String clusterName, String namespaceName) {
        return namespaceBranchAPI.findBranch(appId, env, clusterName, namespaceName);
    }

    public NamespaceBO findBranch(String appId, Env env, String clusterName, String namespaceName) {
        NamespaceDTO namespaceDTO = findBranchBaseInfo(appId, env, clusterName, namespaceName);
        if (namespaceDTO == null) {
            return null;
        }
        return namespaceService.loadNamespaceBO(appId, env, namespaceDTO.getClusterName(), namespaceName);
    }

}

File: apollo-portal/src/main/resources/static/scripts/services/NamespaceBranchService.js
/*
 * Copyright 2021 Apollo Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
appService.service('NamespaceBranchService', ['$resource', '$q', 'AppUtil', function ($resource, $q, AppUtil) {
    var resource = $resource('', {}, {
        find_namespace_branch: {
            method: 'GET',
            isArray: false,
            url: AppUtil.prefixPath() + '/apps/:appId/envs/:env/clusters/:clusterName/namespaces/:namespaceName/branches'
        },
        create_branch: {
            method: 'POST',
            isArray: false,
            url: AppUtil.prefixPath() + '/apps/:appId/envs/:env/clusters/:clusterName/namespaces/:namespaceName/branches'
        },
        delete_branch: {
            method: 'DELETE',
            isArray: false,
            url: AppUtil.prefixPath() + '/apps/:appId/envs/:env/clusters/:clusterName/namespaces/:namespaceName/branches/:branchName'
        },
        merge_and_release_branch: {
            method: 'POST',
            isArray: false,
            url: AppUtil.prefixPath() + '/apps/:appId/envs/:env/clusters/:clusterName/namespaces/:namespaceName/branches/:branchName/merge'
        },
        find_branch_gray_rules: {
            method: 'GET',
            isArray: false,
            url: AppUtil.prefixPath() + '/apps/:appId/envs/:env/clusters/:clusterName/namespaces/:namespaceName/branches/:branchName/rules'
        },
        update_branch_gray_rules: {
            method: 'PUT',
            isArray: false,
            url: AppUtil.prefixPath() + '/apps/:appId/envs/:env/clusters/:clusterName/namespaces/:namespaceName/branches/:branchName/rules'
        }

    });

    function find_namespace_branch(appId, env, clusterName, namespaceName) {
        var d = $q.defer();
        resource.find_namespace_branch({
                appId: appId,
                env: env,
                clusterName: clusterName,
                namespaceName: namespaceName
            },
            function (result) {
                d.resolve(result);
            }, function (result) {
                d.reject(result);
            });
        return d.promise;
    }

    function create_branch(appId, env, clusterName, namespaceName) {
        var d = $q.defer();
        resource.create_branch({
                appId: appId,
                env: env,
                clusterName: clusterName,
                namespaceName: namespaceName
            }, {},
            function (result) {
                d.resolve(result);
            }, function (result) {
                d.reject(result);
            });
        return d.promise;
    }

    function delete_branch(appId, env, clusterName, namespaceName, branchName) {
        var d = $q.defer();
        resource.delete_branch({
                appId: appId,
                env: env,
                clusterName: clusterName,
                namespaceName: namespaceName,
                branchName: branchName
            },
            function (result) {
                d.resolve(result);
            }, function (result) {
                d.reject(result);
            });
        return d.promise;
    }

    function merge_and_release_branch(appId, env, clusterName, namespaceName,
                                      branchName, title, comment, isEmergencyPublish, deleteBranch) {
        var d = $q.defer();
        resource.merge_and_release_branch({
                appId: appId,
                env: env,
                clusterName: clusterName,
                namespaceName: namespaceName,
                branchName: branchName,
                deleteBranch: deleteBranch
            }, {
                releaseTitle: title,
                releaseComment: comment,
                isEmergencyPublish: isEmergencyPublish
            },
            function (result) {
                d.resolve(result);
            }, function (result) {
                d.reject(result);
            });
        return d.promise;
    }

    function find_branch_gray_rules(appId, env, clusterName, namespaceName, branchName) {
        var d = $q.defer();
        resource.find_branch_gray_rules({
                appId: appId,
                env: env,
                clusterName: clusterName,
                namespaceName: namespaceName,
                branchName: branchName
            },
            function (result) {
                d.resolve(result);
            }, function (result) {
                d.reject(result);
            });
        return d.promise;
    }

    function update_branch_gray_rules(appId, env, clusterName,
                                      namespaceName, branchName, newRules) {
        var d = $q.defer();
        resource.update_branch_gray_rules({
                appId: appId,
                env: env,
                clusterName: clusterName,
                namespaceName: namespaceName,
                branchName: branchName
            }, newRules,
            function (result) {
                d.resolve(result);
            }, function (result) {
                d.reject(result);
            });
        return d.promise;
    }

    return {
        findNamespaceBranch: find_namespace_branch,
        createBranch: create_branch,
        deleteBranch: delete_branch,
        mergeAndReleaseBranch: merge_and_release_branch,
        findBranchGrayRules: find_branch_gray_rules,
        updateBranchGrayRules: update_branch_gray_rules
    }
}]);
